[docs] Update limitations page

amandeepmittal opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

Currently, the "Limitations" page has some outdated information, unorganized sections and an inaccurate reference to Expo SDK rather than Expo Go.

# How

<!--
How did you build this feature or fix this bug and why?
-->

This PR 
- puts emphasis on the info around the the EAS Builds
- Links to EAS Build limitations page
- Remove some outdated limitations that are no longer relevant to Expo's ecosystem
- Updates reference of Expo SDK to Expo Go in classic build system limitations
- Updates overall verbiage

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

Changes are tested by running docs locally and visiting http://localhost:3002/introduction/why-not-expo

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

amandeepmittal committed:
[docs] Update limitations page

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -3,108 +3,110 @@ title: Limitations
 import { Collapsible } from '~/ui/components/Collapsible';
 
-Your success will be limited if you don't know the limitations of your tools. A good software engineer strives to understand the tradeoffs in the decisions they make.
+Your success will be limited if you don't know the limitations of your tools. A good software engineer strives to understand the tradeoffs in their decisions.
```
```suggestion
We believe Expo tools are a great option for just about anyone looking to create amazing cross-platform apps quickly. Nonetheless, everyone's project is unique, and there may be reasons Expo is not the right choice for what you're building. Read on for some reasons why you may _not_ want to use Expo tools.
```
The original commentary about what is or isn't a "good developer" feels out-of-character with the voice we've been striving for with docs more recently. Just a suggestion above of a possible voicing that's a little more positive.

amandeepmittal replied:
Thanks for this update! Definitely sounds more positive! 🙏 

I'll just make a small to remove the word "just about" which I'll update on my own.

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -3,108 +3,110 @@ title: Limitations
+Your success will be limited if you don't know the limitations of your tools. A good software engineer strives to understand the tradeoffs in their decisions.
 
-<Collapsible summary="Used Expo before? Here are some recent changes to the managed workflow you should be aware of…">
+## EAS
```
```suggestion
## Expo Application Services
```
It's possible someone can end up here independent of learning anything else about EAS

amandeepmittal committed:
Apply suggestions from code review

Co-authored-by: Keith Kurak <keith.kurak@gmail.com>

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -3,108 +3,110 @@ title: Limitations
-These changes mean Expo can fit many more use cases than before.
+All libraries available to React Native apps are available in Expo projects. Some libraries may require a [Prebuild Config Plugin](/guides/config-plugins/) to be added. The addition of config plugins means:
+
+- Any third-party push notification service, such as OneSignal, is supported.
```
A lot of these, but in particular this one, feel like not a limitation at all, but a previous limitation we've addressed. I think the original intent of this doc was to be the "anti-pitch" that's often at the beginning of a tech talk, a way of building trust with the developer by being open about why this thing they're opinionated in favor of may not be right. for them. So, we may want to avoid saying any _specifically_ positive bullet points in order to avoid diluting the anti-pitch. (generalities like "EAS Build is compatible with all RN projects, **but...**" I think are fair game)

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -3,108 +3,110 @@ title: Limitations
+<Collapsible summary="App size can be as small as you want">
 
-If you're still using `expo build:ios` or `expo build:android`, or using Expo Go for development, it's important to be aware of the limitations that come with those features.
+Apps built with EAS Build can be as small as you want in size since [the entire Expo SDK is no longer bundled](https://expo.fyi/managed-app-size#app-size-with-eas-build) by default.
```
A lot of these, but in particular this one, feel like not a limitation at all, but a previous limitation we've addressed. I think the original intent of this doc was to be the "anti-pitch" that's often at the beginning of a tech talk, a way of building trust with the developer by being open about why this thing they're opinionated in favor of may not be right for them. So, we may want to avoid saying any specifically positive bullet points in order to avoid diluting the anti-pitch. (generalities like "EAS Build is compatible with all RN projects, but..." I think are fair game)

amandeepmittal replied:
I get the idea this is an anti-pitch but we do need a way to convey to the people that the limitations which existed in Expo Go or classic builds aren't true anymore and are resolved with development builds and/or EAS.

keith-kurak replied:
I could see still including a sentence or two calling back to limitations we've addressed with new products- something that makes this point, but without most of the EAS Build anti-pitch actually being about Classic Builds. IMO a link to that future blog post idea about a history of the limitations we've addressed would be even better, .e.g, if you want to read how we've evolved things to address previous limitations, check this out.

amandeepmittal replied:
That sounds good to me. I think after having this discussion with you, I got the fair picture of what we can convert from the Expo Go section. And I totally agree with you that classic build limitation are obsolete and shouldn't be mentioned here.

I'll move the limitations related to classic build system into the archive section.

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -3,108 +3,110 @@ title: Limitations
+EAS Build is a new service and we plan to address many of the current limitations in time. See [EAS Build Limitations](/build-reference/limitations/) for more information on what are the current limitations.
 
-The size for an app built with `expo build` on iOS is approximately 20mb (download), and Android is about 15mb. This is because a bunch of APIs are included regardless of whether or not you are using them &mdash; this lets you publish updates that use new APIs, but comes at the cost of binary size. Some of the APIs that are included are tied to services that you may not be using, for example, the ...
+## Expo Go and the classic `expo build` system
```
I think we should separate out Expo Go and Classic Builds, since the former is sticking around. E.g., "Expo Go is and remains a great way to get started or quickly try something out, but you will need to graduate to dev builds once you use a library with native code that isn't built in."

I'm not even sure we need to mention classic builds in this doc. This is for devs considering Expo, and they shouldn't be jumping into classic builds.

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -3,108 +3,110 @@ title: Limitations
-If you've been reading along each section of the introduction then you will have a pretty good high-level understanding of Expo tools.
+> We are either actively working on or planning to build solutions to all of the limitations listed above, and if you think anything is missing, please bring it to our attention by posting to our [feature requests board](https://expo.canny.io/feature-requests) or the [forums](https://forums.expo.dev/).
+
+If you've been reading along each section of the introduction, you will have a pretty good high-level understanding of Expo tools.
```
```suggestion
```

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -3,108 +3,110 @@ title: Limitations
 ~~To build your app binaries for distribution on the Apple App Store and Google Play Store you will need to follow the same steps that you would in any native project, the Expo build service can't handle it for you. We are working on bringing bare workflow support to the build service in the near future.~~
 
-You can now use [EAS Build](/build/introduction) to build and sign your apps just as easily as in the managed workflow! [Read the announcement blog post](https://blog.expo.dev/expo-application-services-eas-build-and-submit-fc1d1476aa2e).
+You can now use [EAS Build](/build/introduction) to build and sign your apps. [Read the announcement blog post](https://blog.expo.dev/expo-application-services-eas-build-and-submit-fc1d1476aa2e).
```
```suggestion
```
I'd be in favor of striking anything that suggests we just changed this yesterday.

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -3,108 +3,110 @@ title: Limitations
+> If you'd like to continue getting the same development experience that you have with Expo Go, but also use custom native libraries or custom native code, you should check out the [Development builds]()
 
-In the bare workflow, we have full access to the underlying native projects and any native code. It's a "bare" native project with React Native and one or more packages from the Expo SDK installed. Anything that you can do in a native project is possible here.
+As similar to [Development builds], bare workflow also provides access to the underlying native projects and any native code. It's a "bare" native project where developers make direct changes to their native `android` and `ios` project directories rather than continuously generating them on demand using the [Expo config (**app.json**) and prebuild](/workflow/prebuild/).
```
Realizing that I may have suggested so many changes that, if you were to accept them all, not much of this section may be left. In general, I think we could summarize bare workflow under EAS Build as something to the effect of "EAS Build still works if you check in your ios and android native project folders, but prebuild can no longer sync changes from your Expo config to the native projects once you do that; you'll have to configure them on your own with native tools like Xcode and Android Studio."

keith-kurak commented:
One thought that might invalidate a lot of my inline comments- this doc has a lot of vestiges from when EAS was just EAS Build. We might want to organize it more like this doc (https://docs.expo.dev/eas/), with sections for Build, Submit, Update, and Metadata, along with probably Expo Go as the main non-legacy non-EAS-branded thing. Then we could either summarize the limitations of each and/ or link to a doc that goes into more detail, such as is the case for EAS Build, where we already have that.

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -3,108 +3,110 @@ title: Limitations
-<Collapsible summary={<span><span className="strike">Build service only works in the managed workflow</span> (✅ Resolved in December 2020)</span>}>
+<Collapsible summary={<span><span className="strike">Build service only works in the managed workflow</span> (Resolved in December 2020)</span>}>
 
 ~~To build your app binaries for distribution on the Apple App Store and Google Play Store you will need to follow the same steps that you would in any native project, the Expo build service can't handle it for you. We are working on bringing bare workflow support to the build service in the near future.~~
```
Yeah. It feels like too much detail for something that is deprecated and will no longer be available at all in a few months. I did miss the cross out, and I suppose that would at least prevent it from being cherry-picked.

amandeepmittal committed:
Apply suggestions from code review

Co-authored-by: Keith Kurak <keith.kurak@gmail.com>

amandeepmittal committed:
Apply feedback from Keith

amandeepmittal committed:
Remove blockquote

amandeepmittal commented:
@keith-kurak, I've refactored the page in the format you proposed and removed the limitations of classic build system. The page has been trimmed down a lot so if I missed anything, please let me know 🙏 

Also, I think we should move the classic build limitations in the https://docs.expo.dev/archive section until the service is completely shut off. (I can create a follow up PR for that). wdyt?

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -4,107 +4,40 @@ sidebar_title: Limitations
-- Apps built with EAS Build can target children under 13 years old.
+<BoxLink
+  title="EAS Build"
+  description="Some of the missing features from EAS could prevent you from being able to use the service for your applications. Learn more."
```
```suggestion
  description="EAS Build is designed to work for any React Native project, whether or not you also use Expo Prebuild or manage your own native files. In some situations, you may need to change your project configuration, or it may be incompatible with your app. Learn more."
```
Since this will get quoted, want to make sure we're leading with "it should work with just about anything, but check this stuff to be sure". Also not big on the "missing features" verbiage (we may want to update on the EAS Build limitations page, too), since some of the edge cases in there don't really indicate something that's missing, per se.

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -4,107 +4,40 @@ sidebar_title: Limitations
+Similar to [Development builds](/development/introduction), bare workflow also provides access to the underlying native projects and any native code. It's a "bare" native project where you can make direct changes to the project's native **android/** and **ios/** directories rather than continuously generating them on demand using the [Expo config and prebuild](/workflow/prebuild/).
 
-## Limitations of Expo Go and the classic `expo build` system
+EAS Build works with bare workflow projects if you check in the native directories but prebuild can no longer sync changes from the Expo config to the native directories. You'll have to configure the native directories on your own with native tools such as Android Studio or Xcode.
```
```suggestion
EAS Build is compatible with bare workflow projects if you check in the native directories, but no longer runs prebuild, as that could overwrite any manual customizations you've made to the native project files. You'll have to configure the native directories on your own with native tools such as Android Studio or Xcode.
```

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -4,107 +4,40 @@ sidebar_title: Limitations
-Configuring app icons, launch screen, and so on must be configured in the native projects for each platform using the standard native tooling, rather than once using a simple JSON object.
-
-</Collapsible>
+**All of the above have been addressed with EAS Build which is compatible with all type of React Native projects. If you come across them, you can transition your project to use [development builds](/development/introduction/).**
```
```suggestion
**We strongly recommend any projects that may need additional libraries with native code migrate to [development builds](/development/introduction/). It's like a version of Expo Go specifically customized to your app's needs.**
```
Or insert verbiage from your new dev client doc :-). At any rate, I think this should reflect that dev builds are for everyone.

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -4,107 +4,40 @@ sidebar_title: Limitations
 ## Up next
 
-If you've been reading along each section of the introduction then you will have a pretty good high-level understanding of Expo tools.
+We are either actively working on or planning to build solutions to all of the limitations listed above, and if you think anything is missing, please bring it to our attention by posting to our [feature requests board](https://expo.canny.io/feature-requests) or the [forums](https://forums.expo.dev/).
```
```suggestion
We are either actively working on or planning to build solutions to the limitations listed above, and if you think anything is missing, please bring it to our attention by posting to our [feature requests board](https://expo.canny.io/feature-requests) or the [forums](https://forums.expo.dev/).
```
Wanted to couch that statement a little bit, just because some things are limited by their very nature (e.g., Expo Go isn't including custom native code anytime soon)

It might be too much, but wondering if we name-drop Github Discussions as the place for more detailed RfC's/ feature proposals (https://github.com/expo/expo/discussions). Not really committed one way or another on that in this particular spot.

amandeepmittal replied:
I agree on the replacing forums and canny with GitHub discussions. We do not have a special label or category for forums to accept proposals so I guess directing them to GitHub is better. Would you also like the canny link to be removed?

keith-kurak replied:
I think linking to canny and Github for now sounds good. We're still keeping Canny up-to-date at this point

amandeepmittal replied:
Done!

keith-kurak reviewed (commented):
Nice! I think this better focuses on the current state of limitations. Added a few verbiage items.

Since it's a whole new section of content and we probably want to update this doc sooner or later, maybe this should be a separate PR, but I think we should have a callout to limitations in all the core services in there, including EAS Update, Metadata, and Notifications.

Updates in particular would be low-hanging fruit, since I think we can link to a page just like you did for Build

amandeepmittal committed:
Apply suggestions from code review

Co-authored-by: Keith Kurak <keith.kurak@gmail.com>

amandeepmittal commented:
> Since it's a whole new section of content and we probably want to update this doc sooner or later, maybe this should be a separate PR, but I think we should have a callout to limitations in all the core services in there, including EAS Update, Metadata, and Notifications.

We do not have limitations of other core services explicitly defined as for EAS Builds but we do have anti-pitch for EAS metadata (https://docs.expo.dev/eas/metadata/faq/#anti-pitch). Do you think linking it at this stage is okay?

For other services, I'm not sure why there isn't a limitations page but I'd like to discuss that further. Opening a task for it now in linear.

keith-kurak commented:
> 

That EAS Metadata one looks good to link to.

I realize I was thinking of (EAS Update Known Issues)[https://docs.expo.dev/eas/metadata/faq/#anti-pitch] for EAS Update, which needs some significant edits anyway (e.g., not really in preview anymore, Sentry uploads are almost ready, right?). I agree with linking to what we can here and following-up with the rest

amandeepmittal committed:
Add GitHub discussions link and remove forums link

- Also, update verbiage in the **Up next** section

amandeepmittal committed:
Add links to EAS Metadata and EAS Update known issues

amandeepmittal commented:
@keith-kurak, I've added links to FAQ and known issues/anti-pitch for EAS Update and EAS Metadata. Let me know.

> for EAS Update, which needs some significant edits anyway (e.g., not really in preview anymore, Sentry uploads are almost ready, right?).

I wasn't aware of this so thanks for updating me. Have opened a linear task for it (will DM you the link) and will have that conversation from there.

amandeepmittal committed:
Remove EAS Update known issues

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -4,107 +4,49 @@ sidebar_title: Limitations
 ## Up next
 
-If you've been reading along each section of the introduction then you will have a pretty good high-level understanding of Expo tools.
+We are either actively working on or planning to build solutions to all of the limitations listed above, and if you think anything is missing, please bring it to our attention by posting to our [feature requests board](https://expo.canny.io/feature-requests) or [GitHub discussions](https://github.com/expo/expo/discussions).
```
```suggestion
```
I'm struggling to suggest some alternative wording here. I'm uncomfortable saying these limitations are in our roadmap to resolve, because several of them are either impossible in the current mobile ecosystem (e.g., Expo Go limitations), or involve tradeoffs we're unlikely to accept (e.g., some edge cases in EAS Build).

The wording of "if anything is missing" is tricky, too, because the context above is limitations, but those links are for features. If we're missing a limitation on this page, it probably is actually best to post in forums or edit the doc in a PR, which are features built into the bottom of each page already.

Thus, I was wondering, what if we strike this paragraph altogether?

amandeepmittal replied:
I agree and too find this wording irrelevant with the current set of tools and features.

keith-kurak commented on docs/pages/introduction/why-not-expo.mdx:
```diff
@@ -4,107 +4,49 @@ sidebar_title: Limitations
+import { BuildIcon, EasMetadataIcon } from '@expo/styleguide';
 
-Your success will be limited if you don't know the limitations of your tools. A good software engineer strives to understand the tradeoffs in the decisions they make.
+We believe Expo tools are a great option for anyone looking to create amazing cross-platform apps quickly. Nonetheless, everyone's project is unique, and there may be reasons Expo is not the right choice for what you're building. Read on for some reasons why you may **not** want to use Expo tools.
```
```suggestion
We believe Expo tools are a great option for anyone looking to create amazing cross-platform apps quickly. Nonetheless, everyone's project is unique, and there may be reasons Expo is not the right choice for what you're building. Read on for some reasons why you may **not** want to use certain Expo tools with your project.
```
Just thinking of those who suggest Expo is a monolith, that you're either in or out, when really you can use just a little or a lot.

amandeepmittal committed:
Apply suggestions from code review

Co-authored-by: Keith Kurak <keith.kurak@gmail.com>
